executor/aggregate.go

```diff
@@ -45,39 +143,492 @@ type HashAggExec struct {
+		waitGroup.Done()
+	}()
+	sc := ctx.GetSessionVars().StmtCtx
+	if err := w.consumeIntermData(sc); err != nil {
+		w.outputCh <- &AfFinalResult{err: errors.Trace(err)}
+	}
+	w.getFinalResult(sc)
+}
+
 // Next implements the Executor Next interface.
 func (e *HashAggExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.doesUnparallelExec {
+		return errors.Trace(e.unparallelExec(ctx, chk))
+	}
+	return errors.Trace(e.parallelExec(ctx, chk))
+}
+
+func (e *HashAggExec) fetchChildData(ctx context.Context) {
+	var (
+		input *HashAggInput
+		chk   *chunk.Chunk
+		ok    bool
+		err   error
+	)
+	defer func() {
+		for i := range e.partialInputChs {
+			close(e.partialInputChs[i])
+		}
+	}()
+	for {
+		select {
+		case <-e.finishCh:
+			return
+		case input, ok = <-e.inputCh:
+			if !ok {
+				return
+			}
+			chk = input.chk
+		}
+		err = e.children[0].Next(ctx, chk)
+		if err != nil {
+			e.finalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		if chk.NumRows() == 0 {
+			return
+		}
+		input.giveBackCh <- chk
+	}
+}
+
+func (e *HashAggExec) waitPartialWorkerAndCloseOutputChs(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	for _, ch := range e.partialOutputChs {
+		close(ch)
+	}
+}
+
+func (e *HashAggExec) waitFinalWorkerAndCloseFinalOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	close(e.finalOutputCh)
+}
+
+func (e *HashAggExec) prepare4ParallelExec(ctx context.Context) {
+	go e.fetchChildData(ctx)
+
+	partialWorkerWaitGroup := &sync.WaitGroup{}
+	partialWorkerWaitGroup.Add(len(e.partialWorkers))
+	for i := range e.partialWorkers {
+		go e.partialWorkers[i].run(e.ctx, partialWorkerWaitGroup, len(e.finalWorkers))
+	}
+	go e.waitPartialWorkerAndCloseOutputChs(partialWorkerWaitGroup)
+
+	finalWorkerWaitGroup := &sync.WaitGroup{}
+	finalWorkerWaitGroup.Add(len(e.finalWorkers))
+	for i := range e.finalWorkers {
+		go e.finalWorkers[i].run(e.ctx, finalWorkerWaitGroup)
+	}
+	go e.waitFinalWorkerAndCloseFinalOutput(finalWorkerWaitGroup)
+}
+
+// HashAggExec employs one input reader, M partial workers and N final workers to execute parallelly.
+// The parallel execution flow is:
+// 1. input reader reads data from child executor and send them to partial workers
+// 2. partial worker receives the input data, updates the partial results, and shuffle the partial results to the final workers.
+// 3. final worker receives partial results from all the partial workers, evaluates the final results and sends the final results to the main thread.
+func (e *HashAggExec) parallelExec(ctx context.Context, chk *chunk.Chunk) error {
+	if !e.prepared {
+		e.prepare4ParallelExec(ctx)
+		e.prepared = true
+	}
+	for {
+		result, ok := <-e.finalOutputCh
+		if !ok || result.err != nil || result.chk.NumRows() == 0 {
+			if result != nil {
+				return errors.Trace(result.err)
+			}
+			if e.isChildExecReturnEmpty && e.defaultVal != nil {
+				chk.Append(e.defaultVal, 0, 1)
+			}
+			e.isChildExecReturnEmpty = false
```

Comment:
Seems that it not only mean its child have empty data. Does it also mean that `It have returned the  default value`?

Reply:
@winoros 
No
select distinct a, b from t;
all of the agg function are `firstrow`, we would not return default value even the child have empty data.